### DISCUSSION TITLE
Why doesn't pyright checks if types match or not in 'if else' case.

### DISCUSSION THREAD

**Discussion by ensargx** [microsoft/pyright#7676] [Ideas]
Why doesn't pyright check for return types in if-else block. For example in this code;
```python
if '1' == 1:
    ...
```
This if block will never run because if will never return true since right side is int and left side is string.

Why pyright doesn't check for these type of thing? My program wasn't working because i was checking wrong types (return values of founctions, wasn't clear until i checked return types and figure it out) and I think it would be better if pyright did it for me.

First i thought it was a bug that didn't check for return type's of functions but it wasn't even a feature. I just wonder why.

I am using pyright 1.1.345 rn.

**Comment by erictraut** [comment]
Could you provide a more complete code sample that demonstrates the issue? The sample you've provided above doesn't contain any return statements.

Pyright supports a strict-mode diagnostic check called `reportUnnecessaryComparison` that will catch bugs like the one in the code snippet above.

**Reply by ensargx** [reply]
I think the problem was my setup and lack of knowledge in pyright. I didn't knew how to setup enough and now can see what was my mistake.
